field.cpp

```diff
@@ -167,18 +308,138 @@ bool GF2E::operator!=(const GF2E &other) const {
+    // (1, 1, 2, 2, 7, 14, 3), which will have corresponding indexes
+    constexpr uint64_t q_index[u_len - 1] = {0, 0, 1, 1, 4, 5, 2};
+    GF2E b[u_len];
+    b[0] = this->data;
+    for (size_t i = 1; i < u_len; ++i) {
+      GF2E b_p = b[i - 1];
+      GF2E b_q = b[q_index[i - 1]];
+      for (uint64_t m = u[q_index[i - 1]]; m; --m) {
+        b_p = b_p.sqr();
+      }
+      b[i] = reduce_clmul(clmul(b_p.data, b_q.data));
+    }
+    return b[u_len - 1].sqr();
+    break;
+  }
+  case 5: {
+    constexpr uint64_t u[8] = {1, 2, 3, 6, 12, 24, 27, 39};
+    constexpr uint64_t u_len = sizeof(u) / sizeof(u[0]);
+    // q = u[i] - u[i - 1] should give us the corresponding values
+    // (1, 1, 3, 6, 12, 3, 12), which will have corresponding indexes
+    constexpr uint64_t q_index[u_len - 1] = {0, 0, 2, 3, 4, 2, 4};
+    GF2E b[u_len];
+    b[0] = this->data;
+    for (size_t i = 1; i < u_len; ++i) {
+      GF2E b_p = b[i - 1];
+      GF2E b_q = b[q_index[i - 1]];
+      for (uint64_t m = u[q_index[i - 1]]; m; --m) {
+        b_p = b_p.sqr();
+      }
+      b[i] = reduce_clmul(clmul(b_p.data, b_q.data));
+    }
+    return b[u_len - 1].sqr();
+    break;
+  }
+  case 6: {
+    constexpr uint64_t u[9] = {1, 2, 3, 5, 10, 20, 23, 46, 47};
+    constexpr uint64_t u_len = sizeof(u) / sizeof(u[0]);
+    // q = u[i] - u[i - 1] should give us the corresponding values
+    // (1, 1, 2, 5, 10, 3, 23, 1), which will have corresponding indexes
+    constexpr uint64_t q_index[u_len - 1] = {0, 0, 1, 3, 4, 2, 6, 0};
+    GF2E b[u_len];
+    b[0] = this->data;
+    for (size_t i = 1; i < u_len; ++i) {
+      GF2E b_p = b[i - 1];
+      GF2E b_q = b[q_index[i - 1]];
+      for (uint64_t m = u[q_index[i - 1]]; m; --m) {
+        b_p = b_p.sqr();
+      }
+      b[i] = reduce_clmul(clmul(b_p.data, b_q.data));
+    }
+    return b[u_len - 1].sqr();
+    break;
+  }
+  default:
+    return GF2E(this->data);
+    break;
+  }
+}
+
+// This is not faster than calling the operand because we are squaring numbers
+// in GF2_X where X is less than 64
+GF2E GF2E::sqr() const {
+
+  __m128i tmp[2];
+  __m128i res;
+
+  __m128i sqrT = _mm_set_epi64x(0x5554515045444140, 0x1514111005040100);
+  __m128i mask = _mm_set_epi64x(0, 0x0F0F0F0F0F0F0F0F);
+
+  __m128i a = _mm_set_epi64x(0, this->data);
+
+  tmp[0] = _mm_and_si128(a, mask);
+
+  tmp[1] = _mm_srli_epi64(a, 4);
+  tmp[1] = _mm_and_si128(tmp[1], mask);
+
+  tmp[0] = _mm_shuffle_epi8(sqrT, tmp[0]);
+  tmp[1] = _mm_shuffle_epi8(sqrT, tmp[1]);
+
+  res = _mm_unpacklo_epi8(tmp[0], tmp[1]);
+
+  return GF2E(reduce_clmul(res));
+}
+
 void GF2E::to_bytes(uint8_t *out) const {
   uint64_t be_data = htole64(data);
   memcpy(out, (uint8_t *)(&be_data), byte_size);
 }
+
 std::vector<uint8_t> GF2E::to_bytes() const {
   std::vector<uint8_t> buffer(byte_size);
   this->to_bytes(buffer.data());
   return buffer;
 }
 
+uint64_t GF2E::get_data() const { return this->data; }
```

Comment:
Probably don't want this, if you need access to data in a function like eval, make it a friend function.